R/extended-task.R

```diff
@@ -0,0 +1,237 @@
+    #' aware that if the result isn't ready for whatever reason, processing will
+    #' stop in much the same way as `req(FALSE)` does, but when the result is
+    #' ready you'll get invalidated, and when you run again the result should be
+    #' there.
+    #'
+    #' Note that the `result()` method is generally not meant to be used with
+    #' [observeEvent()], [eventReactive()], [bindEvent()], or [isolate()] as the
+    #' invalidation will be ignored.
+    result = function() {
+      switch (private$rv_status(),
+        running = req(FALSE, cancelOutput="progress"),
+        success = if (private$rv_value()$visible) {
+          private$rv_value()$value
+        } else {
+          invisible(private$rv_value()$value)
+        },
+        error = stop(private$rv_error()),
+        # default case (initial, cancelled)
+        req(FALSE)
+      )
+    }
+  ),
+  private = list(
+    func = NULL,
+    # reactive value with "initial"|"running"|"success"|"error"
+    rv_status = NULL,
+    rv_value = NULL,
+    rv_error = NULL,
+    invocation_queue = NULL,
+
+    do_invoke = function(args) {
+      private$rv_status("running")
+      private$rv_value(NULL)
+      private$rv_error(NULL)
+
+      p <- NULL
+      tryCatch({
+        maskReactiveContext({
+          # TODO: Bounce the do.call off of a promise_resolve(), so that the
+          # call to invoke() always returns immediately?
+          result <- do.call(private$func, args)
+          p <- promises::as.promise(result)
+        })
+      }, error = function(e) {
+        private$on_error(e)
+      })
+
+      promises::finally(
+        promises::then(p,
+          onFulfilled = function(value, .visible) {
+            private$on_success(list(value=value, visible=.visible))
+          },
+          onRejected = function(error) {
+            private$on_error(error)
+          }
+        ),
+        onFinally = function() {
+          if (private$invocation_queue$size() > 0) {
+            private$do_invoke(private$invocation_queue$remove())
+          }
+        }
+      )
+
+
+      invisible(NULL)
+    },
+
+    on_error = function(err) {
+      private$rv_status("error")
+      private$rv_error(err)
+    },
+
+    on_success = function(value) {
+      private$rv_status("success")
+      private$rv_value(value)
+    }
+  )
+)
+
+#' Bind `input_task_button` to `ExtendedTask`
+#'
+#' @description
+#' Sets up an [ExtendedTask] to relay its state to an existing
+#' [bslib::input_task_button()], so the task button stays in its "busy" state
+#' for as long as the extended task is running.
+#'
+#' Note that `bind_button_to_task` does _not_ automatically cause button presses
+#' to invoke the extended task; you still need to use [bindEvent()] (or
+#' [observeEvent()]) to cause the button press to trigger
+#' an invocation, as in the example below.
+#'
+#' @export
+bind_button_to_task <- function(task, task_button_id, priority = 1000,
```

Comment:
As a standalone function `bind_button_to_task()` feels very overly coupled with `ExtendedTask` and `input_task_button()` in bslib to be exported from shiny.

One way to look at this is that shiny provides the generic `ExtendedTask` abstract and bslib's task button knows how to extend it. So part of me would rather put this in bslib because it internally it calls bslib functions, while it takes an ExtendedTask object as an input (meaning that it's okay for the task to be created outside of bslib). If moving to bslib, I'd recommend calling it `bind_task_button()` (i.e. `{input,update,bind}_task_button()`) or `bind_task_to_button()`.

```r
adder_task <- 
  shiny::ExtendedTask$new(long_adder_fn) |>
  bslib::bind_task_button("start_task")
```

An alternative: if the implementation stays in shiny, could this be a method on `ExtendedTask`? Or a method plus an argument in `$initialize()`?

```r
# Initialize and bind to button
adder_task <- ExtendedTask$new(long_adder_fn, bind_to_button = "start_task")

# Initialize and bind to the button later
adder_task <- ExtendedTask$new(long_adder_fn)
adder_task$bind_to_button("start_task")

# Chained style
adder_task <- 
  ExtendedTask$
  new(long_adder_fn)$
  bind_to_button("start_task")
```

Reply:
I had essentially the same initial reaction. Feels like the implementation/API should live in Shiny as a part of the `ExtendedTask` API.

Reply:
After discussion, let's go with this:

```r
adder_task <- 
  shiny::ExtendedTask$new(long_adder_fn) |>
  bslib::bind_task_button("start_task")
```

with `bind_task_button` as a generic, and the default implementation throwing an error.